http/src/main/java/io/micronaut/http/body/DefaultMessageBodyHandlerRegistry.java

```diff
@@ -132,8 +133,10 @@ private List<MediaType> resolveMediaTypes(List<MediaType> mediaTypes) {
     @Override
     protected <T> MessageBodyWriter<T> findWriterImpl(Argument<T> type, List<MediaType> mediaTypes) {
         Collection<BeanDefinition<MessageBodyWriter>> beanDefinitions = beanLocator.getBeanDefinitions(
-            Argument.of(MessageBodyWriter.class, type),
-            newMediaTypeQualifier(type, mediaTypes, Produces.class)
+            // Do not put the type here since we are looking for writers that can process the type
+            //      but beanLocator will provide types that can be injected into the searched type
+            Argument.of(MessageBodyWriter.class),
```

Comment:
i dont think this change is correct

Reply:
are you moving this logic to reduce down below? what is the actual change here? I dont understand what youre fixing and why it's fixed

Reply:
The change is the difference between reader and writer: 
* A `MessageBodyReader<T>` would have `T` as a return type therefore for a type `P` we would be looking for any `MessageBodyReader<T>` such that `T extends P`. 
* A `MessageBodyWriter<T>` would have `T` as a method parameter therefore for a type `P` we would be looking for any `MessageBodyWriter<T>` such that `P extends T`.
The main difference is how the `isAssignableFrom` method is called.

Reply:
Yes, and prevent the selection of a body writer for contravariant types, which would throw an error when calling the method.